### DISCUSSION TITLE
liionpack solver 'type' KeyError

### DISCUSSION THREAD

**Discussion by frivollier** [pybamm-team/liionpack#214] [Q&A]
I am looking for help with liionpack  https://github.com/pybamm-team/liionpack

I am getting the following type KeyError while running all examples.  (this is  error from  the example git home page readme.md)

`Traceback (most recent call last):
  File "/Users/frivollier/software_local/BESSTools1/liionpack_DT/test.py", line 46, in <module>
    manager='casadi')
  File "/Users/frivollier/software_local/BESSTools1/venv/lib/python3.7/site-packages/liionpack/solver_utils.py", line 441, in solve
    setup_only=False,
  File "/Users/frivollier/software_local/BESSTools1/venv/lib/python3.7/site-packages/liionpack/solvers.py", line 317, in solve
    self.protocol = lp.generate_protocol_from_experiment(experiment, flatten=True)
  File "/Users/frivollier/software_local/BESSTools1/venv/lib/python3.7/site-packages/liionpack/protocols.py", line 30, in generate_protocol_from_experiment
    typ = op["type"]
KeyError: 'type'
`
running on mac python 3.7

thanks for pointing me in the right direction for some help.

Regards,

Fred

**Comment by frivollier** [comment]
I managed to get it to run on a Windows machine so I have a feeling it is an issue with casadi in a virtual environment on mac. 
I am continuing to debug

**Comment by Saransh-cpp** [comment]
Hi @frivollier

> mac python 3.7

1. The newer versions of PyBaMM support Python 3.8 and 3.9; it is suggested to upgrade to either one of them to avoid installing an older version of PyBaMM.
2. There are some known issues with installing PyBaMM on M1/M2 mac. There are workarounds for this mentioned in https://github.com/pybamm-team/PyBaMM/issues/2346.

Could you also let us know the PyBaMM and liionpack versions you are trying to use? The libraries should work if both of them are on their latest versions -
```
python -m pip show pybamm liionpack
```

**Reply by frivollier** [reply]
I will try to run on python 3.9

here are the versions that I have installed :
Name: pybamm
Version: 22.9
Summary: Python Battery Mathematical Modelling.
Home-page: https://github.com/pybamm-team/PyBaMMAuthor:
Author-email:
License: UNKNOWN
Location: /Users/frivollier/software_local/BESSTools1/venv/lib/python3.7/site-packages
Requires: anytree, autograd, casadi, imageio, julia, jupyter, matplotlib, numpy, pandas, pybtex, scikit-fem, scipy, sympy
Required-by: liionpack
---
Name: liionpack
Version: 0.3.3
Summary: A battery pack simulator for PyBaMM
Home-page:
Author:
Author-email: Tom Tranter <[t.g.tranter@gmail.com](mailto:t.g.tranter@gmail.com)>
License:
Location: /Users/frivollier/software_local/BESSTools1/venv/lib/python3.7/site-packages
Requires: ipython, lcapy, matplotlib, networkx, numpy, openpyxl, pandas, plotly, pybamm, ray, redis, scikit-spatial, scipy, textwrapper, tqdm
Required-by:

**Comment by TomTranter** [comment]
Hi @frivollier I have just released a new verison of liionpack 0.3.5 following PyBaMM's release 23.3. Please let me know if this fixes your issue
